### PR TITLE
ATM Banking System Accounts Management

### PR DESCRIPTION
**Pull request by andreeap00**
### Description

A digital environment for managing ATMs and accounts for a bank is integrated in the current project. 

The code written in Ruby is divided into 3 classes:

1.  main.rb 
   - This file represents the main program, which explores the users' options. 
   - As user, one may choose to:  -  create an account;
                                                      -  display the data of the already created accounts;
                                                      -  retrieve money;
                                                      -  deposit certain sums;  
                                                      -  display current balance;
                                                      -  exit.
     - In order to access money-related data, the personal pieces of information such as name and PIN  of the user are required.
2.  bank-system.rb
   - At the bank-system level, the accounts are managed. Default accounts are created and the possibility for the users to create accounts is ensured.
3.  bank-account.rb
   - At this stage, the functions that allow the users to check current balance, deposit money,  withdraw certain sums and display the account information are ensured.
   - The 5000$ limit per day is set.

### CODE REVIEW

File: bank_account.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+  def deposit_amount(amount)
+    if @pin.nil?
+      puts "PIN is required before insertin deposit."
+    else
+      puts "Insert your PIN."
+      insert_pin = gets.chomp
+      if insert_pin == @pin
+        puts "Are you sure do you want to deposit #{amount}$? \n Press y to continue or n to abort"
+        yes = gets.chomp
```
Comment (VictorMotogna) [inline bank_account.rb]: I think you could some better naming here, like `confirmation_input`.

File: bank_account.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  def withdraw_amount(amount)
+    if @pin.nil?
+      puts "You need to set a PIN before withdrawals."
+    else
+      puts "Insert your PIN: "
+      insert_pin = gets.chomp
+      if insert_pin == @pin
+        puts "Are you sure do you want to withdraw #{amount}$? \n Press y to continue or n to abort"
+        yes = gets.chomp
+        if yes == "y"
```
Comment (VictorMotogna) [inline bank_account.rb]: I noticed that the PIN check & the confirmation part are exactly the same in this method like above. You could extract them to avoid repeating & having something reusable

File: bank_account.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+          if amount <= @remaining_daily_withdrawal_limit && amount > 0  
+            if amount <= @balance
+              @balance -= amount
+              @remaining_daily_withdrawal_limit -= amount
+              @last_withdrawal_date = current_date
+              puts "Withdrawal successful. You withdrew #{amount}$. The current balance is #{@balance}$."
+            else
+              puts "Insufficient funds. Your account balance is #{@balance}$."
+            end
+          elsif amount <= 0
```
Comment (VictorMotogna) [inline bank_account.rb]: You also have this check below, when the user enters the value, before calling the `withdraw_amount` method

File: bank_system.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
+require_relative 'bank_account'
+
+class BankSystem
+  attr_reader :accounts
+
+  def initialize
+    @accounts = []
+
+    # Create default accounts
+    create_account("Naruto Uzumaki", "Software Engineer", "naruto.hokage@yahoo.com", "Emil Racovita Street No 6")
```
Comment (VictorMotogna) [inline bank_system.rb]: I think it would be better to have these in the `main` file

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        if amount <= 0
+          puts "Please enter a positive amount of money to deposit."
+        else
+          account.deposit_amount(amount)
+        end
+      else
+        puts "Incorrect PIN. Deposit failed."
+      end
+    end
+when 5
```
Comment (VictorMotogna) [inline main.rb]: Your indentation is wrong here

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+bank_system = BankSystem.new
+
+puts "Welcome! \n What do you want to do?"
+
+loop do
+  puts "\n Press: \n 1 - create account; \n 2 - display all the accounts; \n 3 - retrieve money; \n 4 - deposit money; \n 5 - display current balance; \n 6 - exit. "
+
+  choice = gets.chomp.to_i
+
+  case choice
```
Comment (VictorMotogna) [inline main.rb]: This method is veery long and hard to follow. I would try to extract the logic into methods like this:
```ruby
case choice
when 1
  create_account
  break
when 2
  bank_system.display_accounts
  break
# .. and so on
```

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+when 5
+    puts "Insert your name."
+    name = gets.chomp
+    account = bank_system.find_account_by_name(name)
+    if account.nil?
+      puts "Account with name #{name} not found. \n Please create an account."
+    else
+      print "Insert PIN: "
+      insert_pin = gets.chomp
+      if insert_pin == account.pin
```
Comment (VictorMotogna) [inline main.rb]: The PIN check from here seems to need the same logic like in the bank account checks. Could you create only 1 method in the Bank Account class to check the PIN & reuse it here as well?

Comment (VictorMotogna) [reply main.rb]: There are also other places that use this that I did not comment on, make sure you replace it everywhere

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    end
+
+    account.set_pin(pin)
+  when 2
+    bank_system.display_accounts
+  when 3
+    puts "\n Insert your name. "
+    name = gets.chomp
+    account = bank_system.find_account_by_name(name)
+    if account.nil?
```
Comment (VictorMotogna) [inline main.rb]: I noticed there are multiple places where you are using
```ruby
if amount.nil?
  puts "error message"
end
```
or other similar conditions like `amount <= 0`

I think the best option for this would be something like an error raise. You can also avoid the error raise in case you would want the program to continue (an error raise would need handling to not stop it).

Or a second option, you could extract logic to validate amount into a method that can be easily reused. You will always check for the same things

File: bank_account.rb [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+  attr_accessor :name, :job, :email, :address, :pin, :balance, :last_withdrawal_date, :remaining_daily_withdrawal_limit
+  
+  WITHDRAWAL_LIMIT_PER_DAY = 5000
+
+  def initialize(name, job, email, address)
+    @name = name
+    @job = job
+    @email = email
+    @address = address
+    @pin = nil
```
Comment (florinionce) [inline bank_account.rb]: you don't need to initialize instance variables with nil, since their default value is nil

File: bank_account.rb [hunk 10]
```diff
@@ -0,0 +1,92 @@
+require 'date'
+
+class BankAccount
+  attr_accessor :name, :job, :email, :address, :pin, :balance, :last_withdrawal_date, :remaining_daily_withdrawal_limit
```
Comment (bogdan-boer) [inline bank_account.rb]: `name, job, email, address` looks like fields that are supposed to be used on an user model than a bank account

File: bank_account.rb [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    @remaining_daily_withdrawal_limit = WITHDRAWAL_LIMIT_PER_DAY
+  end
+
+  def set_pin(pin)
+    @pin = pin
+  end
+
+  def deposit_amount(amount)
+    if @pin.nil?
+      puts "PIN is required before insertin deposit."
```
Comment (bogdan-boer) [inline bank_account.rb]: You should not print things in a class that handles logic, that should be done in the BankSystem

File: bank_account.rb [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    else
+      puts "Insert your PIN."
+      insert_pin = gets.chomp
```
Comment (bogdan-boer) [inline bank_account.rb]: Same for inputs

File: bank_account.rb [hunk 13]
```diff
@@ -0,0 +1,49 @@
+require 'date'
+require_relative './modules/deposit_module'
+require_relative './modules/withdraw_module'
+require_relative './modules/display_module'
+
+class BankAccount
+  include Deposit
+  include Withdraw
+  include InfoDisplay # module with methods used to display info
```
Comment (florinionce) [inline bank_account.rb]: Please don't add this kind of comments, they don't really tell us something. In this case, whenever you include something in ruby it's a module and the module's name is info display. So the code comments itself since we already know that it's a module used to display info.

File: bank_account.rb [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+  
+  attr_accessor :user, :pin, :balance, :last_withdrawal_date, :remaining_daily_withdrawal_limit
+  
+  WITHDRAWAL_LIMIT_PER_DAY = 5000
+
+  def initialize(user)
+    @user = user
+    @pin
```
Comment (florinionce) [inline bank_account.rb:17]: This is also not necessary. in ruby you can use any instance variable without declaring it first (it's just that it will be nil)

File: bank_account.rb [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    @remaining_daily_withdrawal_limit = WITHDRAWAL_LIMIT_PER_DAY
+  end
+
+  # let user set PIN
+  def set_pin(pin)
+    @pin = pin
+  end
+
+  # validate PIN
+ def valid_pin?
```
Comment (florinionce) [inline bank_account.rb]: This is good, but I think nesting code like this is hard to follow. 
```ruby
def valid_pin?
  return false if pin.nil?

 @pin == entered_pin
end
```

Of course, this doesn't print the error messages, but it may be a good idea to print separate the error messages from the validation itself.

File: bank_system.rb [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+class BankSystem
+  attr_reader :accounts
+
+  # array to store bank_account objects
+  def initialize
+    @accounts = []
+  end
+  
+  # create and add bank_account objects
```
Comment (florinionce) [inline bank_system.rb]: same here with the comments

File: main.rb [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+end
+
+def withdraw_amount_interaction(bank_system)
+  puts "\n Insert your name. "
+    name = gets.chomp
+    name.downcase!
+    name.capitalize!
+    account = bank_system.find_account_by_name(name)
+    if account.nil?
+      puts "Account not found. Please create an account first."
```
Comment (florinionce) [inline main.rb]: Same here, I would use early returns instead of nesting the code.

File: main.rb [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+      puts "Not enough funds."
+      end
+    else 
+      puts "Incorrect PIN. Withdrawal failed."
+  end
+end
+
+def deposit_amount_interaction(bank_system)
+  puts "\n Insert your name. "
+    name = gets.chomp
```
Comment (florinionce) [inline main.rb]: the indentation seems wrong

File: spec/bank_system_spec.rb [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+  describe '#initialize' do
+    it 'initialize @accounts' do
+      bank_system = BankSystem.new
+      expect(bank_system.instance_variable_get(:@accounts)).to eq([])
+    end
+  end
+
+  describe '#display_accounts' do
+    it 'display all account-related information' do
+      user1 = User.new("Naruto Uzumaki", "Software Engineer", "naruto.hokage@yahoo.com", "Emil Racovita Street No 6")
```
Comment (florinionce) [inline spec/bank_system_spec.rb]: I would use let instead 
https://www.betterspecs.org/#let

File: spec/bank_system_spec.rb [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+  end
+
+  describe '#find_account_by_name' do
+    it 'get account with corresponding name' do
+      user = User.new("Naruto Uzumaki", "Software Engineer", "naruto.hokage@yahoo.com", "Emil Racovita Street No 6")
+      account = bank_system.create_account(user)
+      found_account = bank_system.find_account_by_name("Naruto Uzumaki")
+      expect(found_account).to eq(account)
+    end
+    it 'nil returned when no account corresponds to name' do
```
Comment (florinionce) [inline spec/bank_system_spec.rb:75]: you missed a new line between blocks :D

**Review by florinionce — APPROVED** [review]
